File: Pod/Classes/ABEIssueManager.m
//
//  ABEIssueManager.m
//  Pods
//
//  Created by Hakon Hanesand on 1/3/16.
//
//

#import "ABEIssueManager.h"

#import "ABEIssue.h"
#import "ABEGithubAPIClient.h"

#import "ABEImgurAPIClient.h"
#import "UIAlertController+ABEErrorAlertController.h"
#import "UIImage+ABEAutoRotation.h"
#import "NSURL+ABERandomImageURL.h"

static double const kABECompressionRatio = 0.5;

@interface ABEIssueManager ()
@property (nonatomic, weak) UIViewController *viewController;
@end

@implementation ABEIssueManager

- (instancetype)initWithReferenceView:(UIView *)referenceView viewController:(UIViewController *)viewController {
    self = [super init];
    
    if (self) {
        _imagesToUpload = [NSMutableArray new];
        _images = [NSMutableArray new];
        _localImageURLs = [NSMutableArray new];
        _issue = [ABEIssue new];
        _viewController = viewController;
        
        [self p_processReferenceView:referenceView];
    }
    
    return self;
}

- (void)p_processReferenceView:(UIView *)referenceView {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        UIGraphicsBeginImageContextWithOptions(referenceView.bounds.size, NO, 0);
        [referenceView drawViewHierarchyInRect:referenceView.bounds afterScreenUpdates:NO];
        UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();

        [self addImageToIssue:image];
    });
}

- (void)addImageToIssue:(UIImage *)image {
    UIImage *flippedImage = [image njh_rotateImageInPreparationForDataConversion];
    NSData *imageData = UIImageJPEGRepresentation(flippedImage, kABECompressionRatio);
    
    [self.images addObject:image];
    [self p_addImageDataToIssue:imageData];
}

- (void)p_addImageDataToIssue:(NSData *)imageData {
    [self willChangeValueForKey:@"imagesToUpload"];
    [self.imagesToUpload addObject:imageData];
    [self didChangeValueForKey:@"imagesToUpload"];
    
    NSURL *saveLocation = [[[NSFileManager defaultManager] URLForDirectory:NSDocumentDirectory inDomain:NSUserDomainMask appropriateForURL:nil create:NO error:nil] njh_URLByAddingRandomImagePathWithExtension:@"jpg"];
    [imageData writeToURL:saveLocation options:kNilOptions error:nil];
    
    [self.localImageURLs addObject:saveLocation];
    
    if (self.completionBlock) {
        self.completionBlock();
    }
    
    __weak typeof(self) weakSelf = self;
    [[ABEImgurAPIClient sharedClient] uploadImageData:imageData success:^(NSString *imageURL) {
        __strong typeof(self) self = weakSelf;
        [self p_didUploadImageWithData:imageData atURL:imageURL];
    } error:^(NSError *error) {
        __strong typeof(self) self = weakSelf;
        [self p_didFailImageUploadWithData:imageData error:error];
    }];
}

- (void)p_didUploadImageWithData:(NSData *)imageData atURL:(NSString *)url {
    NSAssert([self.imagesToUpload containsObject:imageData], @"Images to upload did not contain image that was uploaded.");
    
    [self willChangeValueForKey:@"imagesToUpload"];
    [self.imagesToUpload removeObject:imageData];
    [self didChangeValueForKey:@"imagesToUpload"];
    
    [self.issue attachImageAtURL:url];
}

- (void)p_didFailImageUploadWithData:(NSData *)imageData error:(NSError *)error {
    NSAssert([self.imagesToUpload containsObject:imageData], @"Images to upload did not contain image that just failed uploading.");
    
    [self willChangeValueForKey:@"imagesToUpload"];
    [self.imagesToUpload removeObject:imageData];
    [self didChangeValueForKey:@"imagesToUpload"];
    
    UIAlertController *alertController = [UIAlertController abe_alertControllerFromError:error];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"Retry" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self p_addImageDataToIssue:imageData];
    }]];
    
    [self.viewController presentViewController:alertController animated:YES completion:nil];
}

- (void)saveIssueWithCompletion:(void (^)())completion {
    [[ABEGithubAPIClient sharedClient] saveIssue:self.issue success:^{
        if (completion) {
            completion();
        }
    } error:^(NSError *error) {
        UIAlertController *alertController = [UIAlertController abe_alertControllerFromError:error];
        
        [alertController addAction:[UIAlertAction actionWithTitle:@"Retry" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [self saveIssueWithCompletion:completion];
        }]];
        
        [self.viewController presentViewController:alertController animated:YES completion:nil];
    }];

}

- (NSUInteger)numberOfCurrentlyUploadingImages {
    return self.imagesToUpload.count;
}

- (NSMutableArray *)imagesToUpload {
    return [self mutableArrayValueForKey:[NSString stringWithFormat:@"_%@", NSStringFromSelector(_cmd)]];
}

@end

File: Pod/Classes/ABEissueManager.swift
//
//  ABEissueManager.swift
//  Pods
//
//  Created by Hakon Hanesand on 10/9/16.
//
//

import Foundation
import UIKit

fileprivate extension Array where Element: Equatable {
    
    mutating func removeFirst(element: Element) {
        if let index = self.index(of: element) {
            self.remove(at: index)
        }
    }
}

private let kABECompressionRatio = CGFloat(5.0)

internal protocol ABEIssueManagerDelegate: class {
    
    func issueManagerUploadingStateDidChange(issueManager: ABEIssueManager)
    
    func issueManager(_ issueManager: ABEIssueManager, didFailToUploadImage image: Image, error: IssueReporterError)
    
    func issueManager(_ issueManager: ABEIssueManager, didFailToUploadIssueWithError error: IssueReporterError)
}

internal class ABEIssueManager {
    
    var isUploading: Bool {
        get {
            return images.filter {
                $0.state.contents == .uploading
            }.count > 0
        }
    }
    
    var issue: ABEIssue = ABEIssue()
    
    var images: [Image] {
        get {
            return issue.images
        }
    }
    
    weak var delegate: ABEIssueManagerDelegate?
    
    init(referenceView: UIView, delegate: ABEIssueManagerDelegate? = nil) {
        self.delegate = delegate
        
        drawSnapshotOf(referenceView: referenceView) { [weak self] image in
            self?.add(imageToIssue: image)
        }
    }
    
    private func drawSnapshotOf(referenceView view: UIView, complete: @escaping (UIImage) -> ()) {
        DispatchQueue.global(qos: .userInitiated).async {
            UIGraphicsBeginImageContextWithOptions(view.bounds.size, false, 0)
            view.drawHierarchy(in: view.bounds, afterScreenUpdates: false)
            let image = UIGraphicsGetImageFromCurrentImageContext()!
            UIGraphicsEndImageContext()
            
            complete(image)
        }
    }
    
    func add(imageToIssue image: UIImage) {
        
        let image = Image(image: image)
        self.issue.images.append(image)
        self.persist(image)
    }
    
    func retrySavingImage(image: Image) {
        assert(image.state.contents == .errored, "Can not retry a image that has not errored.")
        persistToCloud(image, withDispatchGroup: DispatchGroup())
    }
    
    private func persist(_ image: Image) {
        
        let group = DispatchGroup()
        let queue = DispatchQueue.global(qos: .`default`)
        
        group.enter()
        
        image.compressionCompletionBlock.contents = { image in
            
            self.persistToDisk(image, withDispatchGroup: group, dispatchQueue: queue)
            self.persistToCloud(image, withDispatchGroup: group)
            
            DispatchQueue.main.async {
                self.delegate?.issueManagerUploadingStateDidChange(issueManager: self)
            }
            
            group.leave()
        }
    }
    
    private func persistToDisk(_ image: Image, withDispatchGroup group: DispatchGroup, dispatchQueue queue: DispatchQueue) {
        
        guard let imageData = image.imageData else {
            print("Image data not set!")
            return
        }
        
        queue.async(group: group) {
            FileManager.write(data: imageData, completion: { url in
                image.localImageURL = url
            }) { error in
                print("Error saving image or screenshot to disk.")
            }
        }
    }
    
    private func persistToCloud(_ image: Image, withDispatchGroup group: DispatchGroup) {
        assert(image.imageData != nil, "Image data must have been set")
        
        group.enter()
        image.state.contents = .uploading
        
        do {
            try ABEImgurAPIClient.shared.upload(imageData: image.imageData!, errorHandler: { [weak self] error in
                group.leave()
                
                image.state.contents = .errored
                
                guard let `self` = self else { return }
                
                DispatchQueue.main.async {
                    self.delegate?.issueManager(self, didFailToUploadImage: image, error: error)
                    self.delegate?.issueManagerUploadingStateDidChange(issueManager: self)
                }
                
            }, success: { [weak self] url in
                group.leave()
                
                guard let `self` = self else { return }
                
                image.cloudImageURL = url
               
                DispatchQueue.main.async {
                    self.delegate?.issueManagerUploadingStateDidChange(issueManager: self)
                }
            })
        } catch {
            print("Error while making image upload request \(error)")
        }
    }
    
    func saveIssue(completion: @escaping () -> ()) {
        do {
            try ABEGithubAPIClient.shared.saveIssue(issue: self.issue, success: completion) { [weak self] error in
                guard let `self` = self else { return }
                
                DispatchQueue.main.async {
                    self.delegate?.issueManager(self, didFailToUploadIssueWithError: error)
                }
            }
        } catch {
            print("Error while trying to save issue : \(error)")
        }
    }
}
